clusterloader/doc/vision.md

```diff
@@ -0,0 +1,369 @@
+
+
+### Measurements
+
+A critical part of Cluster Loader is an ability to check whether tests (defined by
+configs) are satisfying set of Kubernetes performance SLOs.
+Fortunately, for testing a specific functionality, we doesn’t really change the SLO.
+We may want to from time to time tweak how do we measure existing SLOs or introduce
+a new one, but it is fine to require changes to the framework to achieve that.
+
+As a result, mechanisms to measure specific SLOs (or gather other types of metrics)
+will be incorporated into Cluster Loader framework. We will expect that developers
+trying to introduce a new SLO (or change how do we measure the existing one) will
+be modifying that part of Cluster Loader codebase. Within the codebase, we will try
+to provide a relatively easy framework to achieve it though.
+
+At the high level, to implement gathering a given portion of data or measure a new
+SLO, you will need to implement a very simple interface, similar to:
+
+```
+type Measurement interface {
+	StartMeasurement(config MeasurementConfig)
+	StopMeasurement()
+}
+
+struct MeasurementConfig {
+	// Client to access a cluster
+	Clientset k8sclient.ClientSet
+	// More needed parameters to be determined.
+	// TBD: Exact method of passing them. One possibility is having
+	// per-type config and a string field here with json-serialized
+	// config that will be understood only by a given measurement
+	// type.
+}
+```
+
+Providing an implementation of such interface and registering it in the correct
+place will allow you to use those as phases in your config.
+As an example, consider gathering resource usage from system components.
+It will be enough to implement the following methods:
+
+```
+struct ResourceGatherer {
+	// Some fields that you need.
+}
+func (r *ResourceGatherer) StartMeasurement(c MeasurementConfig) {
+	// Initialize gatherer and start gathering metrics.
+}
+funct (r *ResourceGatherer) StopMeasurement() {
+	// Stop gatherer.
+	// Validate and/or report results.
+}
+```
+
+and the registering it in some config file, will let you automatically use:
+
+```
+	ResourceGatherer_Start
+	ResourceGatherer_Stop
+```
+
+as “phases” of your test.
+
+At the beginning, we will provide the following measurement types:
+
+```
+	ResourceGatherer_Start/Stop
+	Profiler_Start/Stop
+	MetricsGatherer_Start/Stop
+	APICallLatencyValidator_Start/Stop
+	PodStartupLatencyValidator_Start/Stop
+```
+
+The remaining question is how to incorporate them into config. To avoid
+introducing another fields to “Phase” type (and making it more confusing to users),
+we introduce a concept of a “test step” where a test step can be either a
+number of phases or a number of measurement actions. We define it as following:
+
+```
+struct Step {
+	// Exactly one of these slices should be non-empty.
+	Phases       []Phase
+	Measurements []string
+}
+```
+
+
+### Top level config
+To define a test scenario, user will be providing the following config:
+
+```
+struct Config {
```

Comment:
This is the biggest question mark, as the goal is to make running these tests accessible to users. While the data structures are robust in allowing CRUD operations as well as multiple steps, the configuration will be nested and elaborate.

The current configuration is much more declarative in nature, as we define:
```
ClusterLoader:
  projects:
    - num: 2
      basename: nginx-explorer
      tuning: default
      RCs:
        - num: 10
          file: nginx.yaml
        - num: 1
          file: explorer-pod.yaml
      pods:
        - num: 50
          image: k8s.gcr.io/pause-amd64:3.0
          basename: pausepods
          file: pod-pause.json
```
This is roughly equivalent to:
```
Config:
  Namespaces: 2
  Steps:
    Phases:
      Creates:	
      - NamespaceRange:
          min: 1
          max: 2
	ObjectType: RC
	ObjectTemplate: /path/nginx.yaml
	ReplicasPerNamespace: 10
	TuningSet: default
      - NamespaceRange:
          min: 1
          max: 2
	ObjectType: RC
	ObjectTemplate: /path/explorer-pod.yaml
	ReplicasPerNamespace: 1
	TuningSet: default
      - NamespaceRange:
          min: 1
          max: 2
	ObjectType: Pod
	ObjectTemplate: /path/pod-pause.json
	ReplicasPerNamespace: 50
	TuningSet: default
```

Reply:
1. With the current definition, it's not possible to test Updates. So we need a concept of phase to achieve that (and that's pretty crucial).

2. Yes - it's roughly that (we may consider building-in the "namespace" deeper into that).
But after thinking much more about it in the meantime, I think that exposing explicitly some resources (e.g. pods, RCs etc. simply doesn't scale). and I want to be it generic so that all resources are treated roughly the same.

Reply:
It could also be possible to add states to the existing config type having a parent slice of projects, that way one can redefine the "state" at various points. Of course there would have to be additional logic to decide of create/update/delete between the states.

Reply:
As discussed during SIG meeting yesterday - I agree that more declarative approach makes sense.
I refined the proposal a bit to accommodate that.

Also, I made it very clear what the logic for deciding create/update/delete should be to make it as simple as possible and understandable to users.